ray_on_golem/server/cluster/sidecars.py

```diff
@@ -0,0 +1,249 @@
+import asyncio
+import logging
+from abc import ABC, abstractmethod
+from asyncio.subprocess import Process
+from datetime import timedelta
+from typing import TYPE_CHECKING, Optional
+
+from golem.utils.asyncio import create_task_with_logging, ensure_cancelled
+from golem.utils.logging import get_trace_id_name
+
+from ray_on_golem.exceptions import RayOnGolemError
+from ray_on_golem.server.mixins import WarningMessagesMixin
+from ray_on_golem.utils import run_subprocess
+
+if TYPE_CHECKING:
+    from ray_on_golem.server.cluster.cluster import Cluster
+    from ray_on_golem.server.cluster.nodes import ClusterNode, HeadClusterNode  # noqa
+
+logger = logging.getLogger(__name__)
+
+
+class ClusterNodeSidecar(WarningMessagesMixin, ABC):
+    @abstractmethod
+    async def start(self) -> None:
+        ...
+
+    @abstractmethod
+    async def stop(self) -> None:
+        ...
+
+    @abstractmethod
+    async def is_running(self) -> bool:
+        ...
+
+
+class MonitorClusterNodeSidecar(ClusterNodeSidecar, ABC):
+    name = "monitor"
+
+    def __init__(self, node: "ClusterNode", cluster: "Cluster"):
+        super().__init__()
+
+        self._node = node
+        self._cluster = cluster
+
+        self._monitor_task: Optional[asyncio.Task] = None
+
+    @abstractmethod
+    async def _monitor(self) -> None:
+        ...
+
+    async def start(self) -> None:
+        if self.is_running():
+            logger.info(f"Not starting `%s` node {self.name}, as it's already running", self._node)
```

Comment:
is this about starting the _node_? or the node's specific monitor?

Reply:
The `self.name` holds the name of a monitor.